stl/inc/ostream

```diff
@@ -1076,6 +1082,260 @@ basic_ostream<_Elem, _Traits>& operator<<(basic_ostream<_Elem, _Traits>& _Ostr,
     // display error code
     return _Ostr << _Errcode.category().name() << ':' << _Errcode.value();
 }
+
+#ifdef __cpp_lib_print
+#ifdef _CPPRTTI
+inline ios_base::iostate _Print_noformat_nonunicode(ostream& _Ostr, const string_view _Str) {
+    // *LOCKED*
+    //
+    // This function is called from a context in which an ostream::sentry has been
+    // constructed.
+    ios_base::iostate _State = ios_base::goodbit;
+
+    _TRY_IO_BEGIN
+    const bool _Was_insertion_successful =
+        (static_cast<size_t>(_Ostr.rdbuf()->sputn(_Str.data(), static_cast<streamsize>(_Str.size()))) == _Str.size());
+    if (!_Was_insertion_successful) [[unlikely]] {
+        _State |= ios_base::badbit;
+    }
+    _CATCH_IO_(ios_base, _Ostr)
+
+    return _State;
+}
+
+template <bool _Add_newline>
+void _Vprint_nonunicode_impl(ostream& _Ostr, const string_view _Fmt_str, const format_args _Fmt_args) {
+    const ostream::sentry _Ok(_Ostr);
+    ios_base::iostate _State = ios_base::goodbit;
+
+    if (!_Ok) [[unlikely]] {
+        _State |= ios_base::badbit;
+    } else [[likely]] {
+        // This is intentionally kept outside of the try/catch block in _Print_noformat_nonunicode()
+        // (see [ostream.formatted.print/3.2]).
+        const string _Output_str = [&]() {
+            if constexpr (_Add_newline) {
+                return _STD vformat(_Ostr.getloc(), _Fmt_str, _Fmt_args) + '\n';
+            } else {
+                return _STD vformat(_Ostr.getloc(), _Fmt_str, _Fmt_args);
+            }
+        }();
+
+        _State |= _STD _Print_noformat_nonunicode(_Ostr, _Output_str);
+    }
+
+    _Ostr.setstate(_State);
+}
+
+inline ios_base::iostate _Print_noformat_unicode(ostream& _Ostr, const string_view _Str) {
+    // *LOCKED*
+    //
+    // This function is called from a context in which an ostream::sentry has been
+    // constructed.
+    ios_base::iostate _State = ios_base::goodbit;
+
+    // The std::ostream& overload of vprint_unicode() expects you to determine whether the stream refers
+    // to a unicode console or not based solely on the std::ostream& provided. That class simply doesn't
+    // provide enough information to know this in every case. The best we can do (without breaking ABI)
+    // is check if the underlying std::streambuf object of the std::ostream& actually refers to a std::filebuf
+    // object. This requires the use of a dynamic_cast. (This is also why the std::ostream& overloads of
+    // std::print() et al. are deleted when RTTI is disabled.)
+    streambuf* const _Streambuf = _Ostr.rdbuf();
+    const auto _Filebuf         = dynamic_cast<filebuf*>(_Streambuf);
+
+    // If we got nullptr, then it probably isn't being used for a unicode console...
+    if (_Filebuf == nullptr) {
+        _State |= _STD _Print_noformat_nonunicode(_Ostr, _Str);
+        return _State;
+    }
+
+    FILE* const _File_stream                         = _Filebuf->_Myfile;

```

Comment:
I don't actually think it's that simple, sadly. The C++ Working Draft states this in [[ostream.formatted.print]/3](https://eel.is/c++draft/ostream.formatted.print#3):

> If the native Unicode API is used, the function flushes `os` before writing `out`. Otherwise (if `os` is not such a stream or the function is `vprint_­nonunicode`), inserts the character sequence [`out.begin()`, `out.end()`) into `os`.

As you mentioned, we need to create an overload for `std::_Print_noformat_nonunicode(std::ostream&)` so that we can insert characters by calling `std::basic_streambuf::sputn()`. The problem, however, is that even after we get a `FILE*` from the `std::filebuf`, we still need to check if that file stream refers to a Unicode console; if it doesn't, then we need to fall back again to the specialized version of `std::_Print_noformat_nonunicode(std::ostream&)`.

This leaves only two actions which can be re-used: flushing the file stream (via `std::fflush()`, which is called internally by `std::basic_filebuf::pubsync()`) and calling `__std_print_to_unicode_console()`. We could do these in a separate private function (maybe something like `std::_Print_unicode_console()`) and use that in both versions of `std::_Print_noformat_unicode()` as applicable.